File: 01-Languages/Go/07-Collections/Solutions/main.go
// main.go - word frequency counting with a map, ranked via sort.Slice.
package main

import (
	"fmt"
	"sort"
	"strings"
)

type wordCount struct {
	word  string
	count int
}

func wordFrequency(text string) map[string]int {
	cleaned := strings.ToLower(text)
	cleaned = strings.NewReplacer(".", "", ",", "", "!", "", "?", "").Replace(cleaned)
	freq := make(map[string]int)
	for _, word := range strings.Fields(cleaned) {
		freq[word]++
	}
	return freq
}

func topN(freq map[string]int, n int) []wordCount {
	entries := make([]wordCount, 0, len(freq))
	for word, count := range freq {
		entries = append(entries, wordCount{word, count})
	}
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].count != entries[j].count {
			return entries[i].count > entries[j].count
		}
		return entries[i].word < entries[j].word
	})
	if len(entries) > n {
		entries = entries[:n]
	}
	return entries
}

func main() {
	freq := wordFrequency("Cats, cats, and dogs. Dogs love cats!")
	for _, entry := range topN(freq, 2) {
		fmt.Printf("%s: %d\n", entry.word, entry.count)
	}
}

File: 01-Languages/Rust/07-Collections/Solutions/main.rs
// main.rs - word frequency counting with HashMap, ranked via iterator adapters and sort_by.

use std::collections::HashMap;

fn word_frequency(text: &str) -> HashMap<String, i32> {
    let cleaned: String = text
        .to_lowercase()
        .chars()
        .filter(|c| !".,!?".contains(*c))
        .collect();

    let mut freq = HashMap::new();
    for word in cleaned.split_whitespace() {
        *freq.entry(word.to_string()).or_insert(0) += 1;
    }
    freq
}

fn top_n(freq: &HashMap<String, i32>, n: usize) -> Vec<(String, i32)> {
    let mut entries: Vec<(String, i32)> = freq.iter().map(|(k, v)| (k.clone(), *v)).collect();
    entries.sort_by(|a, b| b.1.cmp(&a.1).then_with(|| a.0.cmp(&b.0)));
    entries.truncate(n);
    entries
}

fn main() {
    let freq = word_frequency("Cats, cats, and dogs. Dogs love cats!");
    for (word, count) in top_n(&freq, 2) {
        println!("{}: {}", word, count);
    }
}
